### PR TITLE
[dynamo] add some cpython debugging methods

### PR DESCRIPTION
**Pull request by williamwen42**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #138284
* __->__ #138030

This PR enables you to inspect PyObjects in C using `INSPECT(...)` without requiring https://docs.python.org/3/howto/gdb_helpers.html. `torch._dynamo.eval_frame.raise_sigtrap` can also be used to set gdb breakpoints while running Python code, e.g.

```python
x = x + 1
torch._dynamo.eval_frame.raise_sigtrap();
# can breakpoint on ceval.c:CALL to breakpoint the `sin` call in C.
x = torch.sin(x)
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

File: torch/csrc/dynamo/debug_macros.h [hunk 1]
```diff
@@ -53,6 +53,25 @@ extern "C" {
 
 #endif
 
+// e.g. INSPECT(obj1, obj2)
+// in order to inspect obj1, obj2 at the Python level, in pdb.
+// Alias for torch._dynamo.utils._breakpoint_for_c_dynamo(...)
+// WARNING: makes a Python function call! Make sure eval frame callback is
+// unset!
```
Comment (StrongerXi) [inline torch/csrc/dynamo/debug_macros.h:60]: > makes a Python function call! Make sure eval frame callback is unset!

Any chance we could automate this in the macro, like adding some wrapper calls to unset and reset the callback?

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**